tutorials-v5/miscellaneous/excitation-number-restricted-states-jc-chain.md

```diff
@@ -0,0 +1,264 @@
+    Hint_cc = 0
+    for n in range(N - 1):
+        Hint_cc += 0.9 * (a[n].dag() * a[n + 1] + a[n + 1].dag() * a[n])
+
+    H = H0 + Hint_ac + Hint_cc
+
+    e_ops = [x.dag() * x for x in d]
+    c_ops = [0.01 * x for x in a]
+
+    times = np.linspace(0, 250, 1000)
+    L = liouvillian(H, c_ops)
+    result = mesolve(
+        H, psi0, times, c_ops, e_ops, options={"nsteps": 5000, "store_states": True}
+    )
+    return result, H, L
+```
+
+### Regular QuTiP States and Operators
+
+```python
+d = [
+    tensor(
+        [
+            destroy(dim1) if idx1 == idx2 else identity(dim1)
+            for idx1, dim1 in enumerate(dims)
+        ]
+    )
+    for idx2, _ in enumerate(dims)
+]
+psi0 = tensor(
+    [
+        basis(dim, init_excite) if idx == 1 else basis(dim, 0)
+        for idx, dim in enumerate(dims)
+    ]
+)
+```
+
+```python
+%time res1, H1, L1 = solve(d, psi0)
+```
+
+### Using ENR States and Operators
+
+```python
+d = enr_destroy(dims, excite)
+psi0 = enr_fock(dims, excite, [init_excite if n == 1 else 0 for n in range(2 * N)])
+```
+
+```python
+%time res2, H2, L2 = solve(d, psi0)
+```
+
+### Comparison of Expectation Values
+
+```python
+fig, axes = plot_expectation_values([res1, res2])
+fig.set_figwidth(10)
+fig.set_figheight(8)
+for idx, ax in enumerate(axes):
+    if idx % 2:
+        ax.set_ylabel(f"Atom {idx//2}")
+    else:
+        ax.set_ylabel(f"Cavity {idx//2}")
+    ax.set_ylim(-0.1, 1.1)
+    ax.grid()
+fig.tight_layout()
+```
+
+### Calculation of Partial Trace
+
+The usage of ENR states makes many standard QuTiP features fail.
+```ptrace``` is one of those.
+Below we demonstrate how the partial trace for ENR states can be calculated and show the corresponding result together with the standrad QuTiP approach.
+
+```python
+def ENR_ptrace(rho, sel, excitations):
+    if isinstance(sel, int):
+        sel = np.arrau([sel])
+    else:
+        sel = np.asarray(sel)
+
+    if (sel < 0).any() or (sel >= len(rho.dims[0])).any():
+        raise TypeError("Invalid selection index in ptrace.")
+
+    drho = rho.dims[0]
+    _, state2idx, _ = enr_state_dictionaries(drho, excitations)
+
+    dims_short = np.asarray(drho).take(sel).tolist()
+    nstates2, state2idx2, _ = enr_state_dictionaries(dims_short, excitations)
+
+    # construct new density matrix
+    rhout = np.zeros((nstates2, nstates2), dtype=np.complex64)
+    # dimensions of traced out system
+    rest = np.setdiff1d(np.arange(len(drho)), sel)
+    for state in state2idx:
+        for state2 in state2idx:
+            # if the parts of the states of the systems(s) being traced out are diagonal, add this to the new DM
+            if np.all(np.asarray(state).take(rest) == np.asarray(state2).take(rest)):
+                rhout[
+                    state2idx2[tuple(np.asarray(state).take(sel))],
+                    state2idx2[tuple(np.asarray(state2).take(sel))],
+                ] += rho[state2idx[state], state2idx[state2]]
+
+    return Qobj(rhout)
```

Comment:
would be nice to set the dims, using EnrSpace()?

Reply:
As far as I can see the *EnrSpace* class is not available to import. In the *energy_restricted.py* it is not part of __all__. I can make a pull request there but it will take some time nontheless.

Reply:
i think you can manually grab it with `from qutip.core.energy_restricted import EnrSpace`